src/core/address_payload_unification_design.md

```diff
@@ -0,0 +1,556 @@
+    instance for any given node: that is, it will try to call
+    `createReference` only once. Plugins shouldn’t depend on this for
+    correctness, but it might be helpful for their caching patterns.
+  - The new serialized form will attach plugin names to nodes, either
+    individually or by bucketing (each has advantages). Nodes themselves
+    will be serialized `NodePayload`s. At `fromJSON` time, the graph
+    will re-enrich all nodes with their appropriate plugins to get
+    payloads and references.
+
+Additionally, the proper semantics of `nodeReference` are not entirely
+clear. There are four cases, depending on the `address` argument passed
+to this method:
+
+  1. The address corresponds to a node in the graph.
+  2. The address corresponds to a node that does not appear in the
+     graph, but has incident edges in the graph.
+  3. The address corresponds to a phantom node—a node that does not
+     exist in the graph and does not have incident edges, but still
+     appears in `_nodes` due to a prior call to `removeNode`.
+  4. The address has never been seen before.
+
+Correspondingly:
+
+  - The semantics for case (1) are clear.
+  - It seems highly probable that, for case (2), we should return a
+    non-null reference: otherwise, it is not clear how we can explore
+    the structure around nodes that we know exist but do not have
+    directly (e.g., to answer the question, “what other commits has the
+    author of this commit authored?” when the “commit” notion comes from
+    a different plugin than the “author” notion (say, a Git plugin and
+    an OAuth plugin), and only the commit graph is loaded).
+  - It also seems clear that cases (3) and (4) should have the same
+    semantics as each other, but it is less clear what these should be.
+  - In these cases, we could elect to either return `undefined` or to
+    return a “null object”: an object for which `get()` returns `null`
+    and `neighbors(_)` returns `[].values()`.
+  - We should also ask—is the address required to be associated with a
+    known plugin? If so, should we ask the plugin to give us a
+    `NodeReference` instance specific to that plugin?
+  - If we later learn about this node, can the formerly dangling
+    reference come “back online”? Should it?
+  - Does this mean that calling `nodeReference` can change the internal
+    state of the graph? How does this impact the graph’s ability to
+    compress its `_nodes`?
+
+Some of these latter questions apply in case (2), too.
+
+One resolution to, I think, all of these questions is to change the
+`PluginHandler.createReference` method so that it does not accept a
+`NodePayload`, or at least to make the payload optional. This has the
+nice property that `PluginHandler<NR, NP>` has two simple functions: one
+to enrich a base reference into an `NR`, and one to enrich a JSON blob
+into an `NP`. I suspect that this is likely a good solution, but I am
+interested to hear other opinions.
```

Comment:
I believe that requesting a reference to an address should always return a valid NodeReference (albeit one which might have no payload and no neighbors). Since we are able to provide valid implementations of all the NR functions even if it is not defined. Implementing inconsistent behavior where we don't return those perfectly valid NRs seems needlessly confusing to me. Your proposal on how to do this in a manner that is consistent with plugins makes sense to me.

Reply:
1. seems totally reasonable to me.

I don't think 2. and 3. are necessary implications. When the graph gets a request for a NodeReference that does not exist in the graph, it sets its index to null. Whenever `_ensureUpToDate()` is called on that `NodeReference`, it checks its address in the graph to see if it has an index. If it does, then it can delegate to the graph to find its neighbors. If not, then it knows it has no empty and it does not have a corresponding payload. I believe this behavior is correct, and also these "phantom references" can be GC'd like any other object, are not pointed to by the graph, and generating them does not require polluting the graph's internal state.

Reply:
I've adopted this and written the implications in the doc.

Reply:
(Didn’t see your last two comments at time of last push.)

It sounds like what you’re suggesting requires abandoning the goal that
we only create a `NodeReference` once per address. This is totally
reasonable, in my opinion.

It sounds like a consequence of what you’re suggesting is that we need
to bump the `_indexVersion` whenever a node is added to the graph.
Consider the following scenario:

```javascript
const g = new Graph();
const ref = g.nodeReference(anAddress);
// now, `ref`'s internal `index` is `null`
assert(g.get() == null);
g.addNode({address: anAddress, payload: {}});
assert(g.get() != null);
```

For the second assertion to succeed, the cache must have been
invalidated by the `addNode`. Or am I misunderstanding something?